conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java

```diff
@@ -0,0 +1,99 @@
+import io.undertow.util.AttachmentKey;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.xnio.XnioExecutor;
+
+final class ConjureAsyncRequestProcessing implements AsyncRequestProcessing {
+
+    private static final Executor DIRECT_EXECUTOR = MoreExecutors.directExecutor();
+    private static final AttachmentKey<ListenableFuture<?>> FUTURE = AttachmentKey.create(ListenableFuture.class);
+    // If the request is ended before the future has completed, cancel the future to signal that work
+    // should be stopped. This occurs when clients cancel requests or connections are closed.
+    private static final ExchangeCompletionListener COMPLETION_LISTENER = (exchange, nextListener) -> {
+        ListenableFuture<?> future = exchange.getAttachment(FUTURE);
+        if (future != null && !future.isDone()) {
+            future.cancel(false);
+        }
+        nextListener.proceed();
+    };
+
+    private final Serializer<SerializableError> exceptionSerializer = ConjureExceptions.serializer();
+
+    private final Duration timeout;
+
+    ConjureAsyncRequestProcessing(Duration timeout) {
+        this.timeout = Preconditions.checkNotNull(timeout, "Timeout is required");
+    }
+
+    @Override
+    public <T> void register(
+            ListenableFuture<T> future,
+            ReturnValueWriter<T> returnValueWriter,
+            HttpServerExchange exchange) throws IOException {
+        Preconditions.checkNotNull(future, "future");
+        Preconditions.checkNotNull(returnValueWriter, "returnValueWriter");
+        Preconditions.checkNotNull(exchange, "exchange");
+
+        // TODO(ckozak): potential optimization: avoid both dispatching and scheduling a timeout if the future is
+        // already complete.
+
+        // Attach data to the exchange in order to reuse a stateless listener
+        exchange.putAttachment(FUTURE, future);
+        // Cancel the future if the exchange is completed before the future is done
+        exchange.addExchangeCompleteListener(COMPLETION_LISTENER);
+
+        XnioExecutor.Key timeoutKey = exchange.getIoThread()
+                .executeAfter(() -> future.cancel(false), timeout.toMillis(), TimeUnit.MILLISECONDS);
+        future.addListener(timeoutKey::remove, DIRECT_EXECUTOR);
+        // Dispatch the registration task. We're executing inside of root handler, so the dispatched task will
+        // execute on this thread after the current handler completes execution. This allows us to avoid racing
+        // registration with completion.
+        // This leaves the exchange in a dispatched state preventing the calling Connectors.executeRootHandler
+        // from terminating this request while the future awaits completion.
+        exchange.dispatch(() -> Futures.addCallback(future, new FutureCallback<T>() {
```

Comment:
the javadoc indicates that you should check whether `isInIoThread`

Reply:
I will add class level javadoc describing the thread model. The description in HttpServerExchange.dispatch javadoc describes how to move work from an I/O thread to a task thread, but that's not what we want to accomplish here -- we already know we're on a task thread, but we need to mark the exchange as dispatched while we wait for the future to complete. I updated the comment, hopefully that makes the intent clearer.